### DISCUSSION TITLE
Best way to support limiting queue concurrency?

### DISCUSSION THREAD

**Discussion by jordan-brough** [sidekiq/sidekiq#5266] [Q&A] [answered]
Is [this](https://github.com/mperham/sidekiq/wiki/Best-Practices#3-embrace-concurrency):

> targeting a sidekiq process at a queue with a defined number of threads

still the best way to limit concurrency in Sidekiq? We're finding that costly and we're wondering if there are solutions we aren't aware of.

### Context

We have a multi-tenant product that builds on top of **slow** and **unreliable** tenant systems that have **bursty** workloads. We make a lot of API calls to those systems and when possible we do that asynchronously via Sidekiq to help deal with slowness/glitches/outages.

Sidekiq is great, but we still need to:
1) Rate-limit the number of API calls we perform concurrently
2) Prevent an outage or slowdown on one tenant from affecting other tenants.

### Problems/Solutions

* [Sidekiq::Limiter](https://github.com/mperham/sidekiq/wiki/Ent-Rate-Limiting) helps with item 1, but a slowdown on a single tenant's API can still hurt item 2. I.e. [this problem](https://github.com/mperham/sidekiq/wiki/Ent-Rate-Limiting#limiting-is-not-throttling):

    > The concurrent limiter does not limit the number of jobs that will be executed. If you have 20 worker threads available, 20 jobs will be taken from the queue and executed. However, if you have a concurrent limiter set to 5 concurrent jobs, for example, only 5 of those 20 jobs will be able to execute the code inside the within_limit block. The other 15 will wait for up to wait_timeout seconds for a lock to become available before erroring and retrying.

* The "multiple single-queue thread-limited Sidekiq workers" approach should solve both items 1 and 2, but it costs a lot more to run a lot of workers than a lot of threads, especially when the workload per tenant is bursty.

Are there any other Sidekiq-supported solutions for this type of situation that I'm not aware of?

------------------

FYI another option we're considering is to use separate SQS queues per tenant that trigger a Lambda function, because SQS queues don't incur costs (only sending/receiving messages does) and on the worker side we'd only pay per message (per Lambda invocation) instead of paying per hour for lots of workers that need to be constantly polling for jobs.  We'd prefer to use Sidekiq though.

**Comment by mperham** [comment, marked as answer]
I don't have any other solutions. Sidekiq, like any tool, has advantages and disadvantages. Dynamic queue throttling is an advanced usecase without an easy or simple solution. My Faktory Enterprise product does have native static queue throttling (https://github.com/contribsys/faktory/wiki/Ent-Throttling) but does not support dynamic queue throttling (i.e. per-user). Please also note queue sharding, another technique to throttle overactive tenants.

https://www.mikeperham.com/2019/12/17/workload-isolation-with-queue-sharding/

As you point out, there are cases where SQS might work better. Only you can determine if a hybrid approach would work best for your system.

**Reply by jordan-brough** [reply]
Thank you for the quick response @mperham!

In this case, given our high rate of failure in the APIs we deal with, the bursty nature of our jobs, and the level of throttling we need to do (e.g. to not overload the 3rd party APIs), I think we'll end up with a simpler and more monitorable/debuggable system by going with something that's more isolated per tenant, as long as we can do that in a cost-effective way.  E.g. we'd like to be able to easily monitor latency per tenant, which seems difficult with queue sharding. Also in the default sharding scheme described it takes just 2 tenant outages to knock out a third tenant — and increasing that "2" to e.g. 3 or 4 starts to quickly decrease how many tenants you can service with a given number of workers.

I guess we'd like to share queue storage (e.g. Redis) and share workers, but without having to pay for a full Rails process to poll each tenant, and without overloading Redis. I'm thinking an SQS FIFO queue for each tenant + Lambda triggers will end up being substantially cheaper since it is pay-per-usage without requiring polling, while still being very monitorable and debuggable, and will allow us to limit concurrency per queue (FIFO queues will prevent more jobs from going out until the currently-processing jobs are finished).  If Sidekiq ever adds functionality that hits those points we'd love to hear about it!

In any case though, thank you for the link on queue sharding — that looks very promising and I think we'll use that with Sidekiq in some others areas in our product!
